### ISSUE TITLE
Cannot build nor run with python-urllib3 1.21

### ISSUE THREAD

**Issue by protist** [pimutils/vdirsyncer#627]
Before you submit bug reports: https://vdirsyncer.pimutils.org/en/stable/contributing.html

Things to include in your bugreport:

* Your vdirsyncer version: 0.15.0
* If applicable, which server software (and which version) you're using: Nextcloud 11.0.3
* Your Python version: 3.6.1
* Your operating system: Arch Linux
* Your config file:
```
[general]
status_path = "~/.vdirsyncer/status/"
[pair nextcloud_contacts]
a = "nextcloud_contacts_local"
b = "nextcloud_contacts_remote"
collections = ["from a", "from b"]
conflict_resolution = ["command", "vimdiff"]
[storage nextcloud_contacts_local]
type = "filesystem"
path = "~/.contacts/"
fileext = ".vcf"
[storage nextcloud_contacts_remote]
type = "carddav"
url = "https://foo.dynu.com/nextcloud/remote.php/carddav/"
useragent = "Mozilla"
verify_fingerprint = "<REMOVED>"
verify = false
username = "protist"
password.fetch = ["command", "keyring", "get", "owncloud", "protist"]
```
* Use `vdirsyncer -vdebug` for debug output. The output is sensitive, but
  please attach at least the last few lines before the error (if applicable),
  censored as necessary. This is almost always the most useful information.

There appears to be another issue with python-urllib3 (as per #518). If I install python-urllib3 1.20, vdirsyncer builds fine. If I install python-urllib3 1.21, it fails the tests when I try to build.
```
...
tests/system/utils/test_main.py ..FF.
...
==================================================================================================== FAILURES ====================================================================================================
___________________________________________________________ test_request_ssl_fingerprints[94:FD:7A:CB:50:75:A4:69:82:0A:F8:23:DF:07:FC:69:3E:CD:90:CA] ___________________________________________________________

httpsserver = <SecureContentServer(<class 'pytest_localserver.https.SecureContentServer'>, started 139956190172928)>, fingerprint = '94:FD:7A:CB:50:75:A4:69:82:0A:F8:23:DF:07:FC:69:3E:CD:90:CA'

    @pytest.mark.skipif(_fingerprints_broken(),
                        reason='https://github.com/shazow/urllib3/issues/529')
    @pytest.mark.parametrize('fingerprint', [
        '94:FD:7A:CB:50:75:A4:69:82:0A:F8:23:DF:07:FC:69:3E:CD:90:CA',
        '19:90:F7:23:94:F2:EF:AB:2B:64:2D:57:3D:25:95:2D'
    ])
    def test_request_ssl_fingerprints(httpsserver, fingerprint):
        httpsserver.serve_content('')  # we need to serve something
    
        http.request('GET', httpsserver.url, verify=False,
>                    verify_fingerprint=fingerprint)

tests/system/utils/test_main.py:54: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
vdirsyncer/http.py:162: in request
    r = func(method, url, **kwargs)
/usr/lib/python3.6/site-packages/requests/sessions.py:488: in request
    resp = self.send(prep, **send_kwargs)
/usr/lib/python3.6/site-packages/requests/sessions.py:609: in send
    r = adapter.send(request, **kwargs)
/usr/lib/python3.6/site-packages/requests/adapters.py:390: in send
    conn = self.get_connection(request.url, proxies)
/usr/lib/python3.6/site-packages/requests/adapters.py:296: in get_connection
    conn = self.poolmanager.connection_from_url(url)
/usr/lib/python3.6/site-packages/urllib3/poolmanager.py:271: in connection_from_url
    pool_kwargs=pool_kwargs)
/usr/lib/python3.6/site-packages/urllib3/poolmanager.py:219: in connection_from_host
    return self.connection_from_context(request_context)
/usr/lib/python3.6/site-packages/urllib3/poolmanager.py:230: in connection_from_context
    pool_key = pool_key_constructor(request_context)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

key_class = <class 'requests.packages.urllib3.poolmanager.PoolKey'>
request_context = {'assert_fingerprint': '94:FD:7A:CB:50:75:A4:69:82:0A:F8:23:DF:07:FC:69:3E:CD:90:CA', 'block': False, 'host': '127.0.0.1', 'maxsize': 10, ...}

    def _default_key_normalizer(key_class, request_context):
        """
        Create a pool key out of a request context dictionary.
    
        According to RFC 3986, both the scheme and host are case-insensitive.
        Therefore, this function normalizes both before constructing the pool
        key for an HTTPS request. If you wish to change this behaviour, provide
        alternate callables to ``key_fn_by_scheme``.
    
        :param key_class:
            The class to use when constructing the key. This should be a namedtuple
            with the ``scheme`` and ``host`` keys at a minimum.
        :type  key_class: namedtuple
        :param request_context:
            A dictionary-like object that contain the context for a request.
        :type  request_context: dict
    
        :return: A namedtuple that can be used as a connection pool key.
        :rtype:  PoolKey
        """
        # Since we mutate the dictionary, make a copy first
        context = request_context.copy()
        context['scheme'] = context['scheme'].lower()
        context['host'] = context['host'].lower()
    
        # These are both dictionaries and need to be transformed into frozensets
        for key in ('headers', '_proxy_headers', '_socks_options'):
            if key in context and context[key] is not None:
                context[key] = frozenset(context[key].items())
    
        # Map the kwargs to the names in the namedtuple - this is necessary since
        # namedtuples can't have fields starting with '_'.
        for key in list(context.keys()):
            context['key_' + key] = context.pop(key)
    
        # Default to ``None`` for keys missing from the context
        for field in key_class._fields:
            if field not in context:
                context[field] = None
    
>       return key_class(**context)
E       TypeError: __new__() got an unexpected keyword argument 'key_assert_fingerprint'

/usr/lib/python3.6/site-packages/urllib3/poolmanager.py:95: TypeError
_________________________________________________________________ test_request_ssl_fingerprints[19:90:F7:23:94:F2:EF:AB:2B:64:2D:57:3D:25:95:2D] _________________________________________________________________

httpsserver = <SecureContentServer(<class 'pytest_localserver.https.SecureContentServer'>, started 139956190172928)>, fingerprint = '19:90:F7:23:94:F2:EF:AB:2B:64:2D:57:3D:25:95:2D'

    @pytest.mark.skipif(_fingerprints_broken(),
                        reason='https://github.com/shazow/urllib3/issues/529')
    @pytest.mark.parametrize('fingerprint', [
        '94:FD:7A:CB:50:75:A4:69:82:0A:F8:23:DF:07:FC:69:3E:CD:90:CA',
        '19:90:F7:23:94:F2:EF:AB:2B:64:2D:57:3D:25:95:2D'
    ])
    def test_request_ssl_fingerprints(httpsserver, fingerprint):
        httpsserver.serve_content('')  # we need to serve something
    
        http.request('GET', httpsserver.url, verify=False,
>                    verify_fingerprint=fingerprint)

tests/system/utils/test_main.py:54: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
vdirsyncer/http.py:162: in request
    r = func(method, url, **kwargs)
/usr/lib/python3.6/site-packages/requests/sessions.py:488: in request
    resp = self.send(prep, **send_kwargs)
/usr/lib/python3.6/site-packages/requests/sessions.py:609: in send
    r = adapter.send(request, **kwargs)
/usr/lib/python3.6/site-packages/requests/adapters.py:390: in send
    conn = self.get_connection(request.url, proxies)
/usr/lib/python3.6/site-packages/requests/adapters.py:296: in get_connection
    conn = self.poolmanager.connection_from_url(url)
/usr/lib/python3.6/site-packages/urllib3/poolmanager.py:271: in connection_from_url
    pool_kwargs=pool_kwargs)
/usr/lib/python3.6/site-packages/urllib3/poolmanager.py:219: in connection_from_host
    return self.connection_from_context(request_context)
/usr/lib/python3.6/site-packages/urllib3/poolmanager.py:230: in connection_from_context
    pool_key = pool_key_constructor(request_context)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

key_class = <class 'requests.packages.urllib3.poolmanager.PoolKey'>
request_context = {'assert_fingerprint': '19:90:F7:23:94:F2:EF:AB:2B:64:2D:57:3D:25:95:2D', 'block': False, 'host': '127.0.0.1', 'maxsize': 10, ...}

    def _default_key_normalizer(key_class, request_context):
        """
        Create a pool key out of a request context dictionary.
    
        According to RFC 3986, both the scheme and host are case-insensitive.
        Therefore, this function normalizes both before constructing the pool
        key for an HTTPS request. If you wish to change this behaviour, provide
        alternate callables to ``key_fn_by_scheme``.
    
        :param key_class:
            The class to use when constructing the key. This should be a namedtuple
            with the ``scheme`` and ``host`` keys at a minimum.
        :type  key_class: namedtuple
        :param request_context:
            A dictionary-like object that contain the context for a request.
        :type  request_context: dict
    
        :return: A namedtuple that can be used as a connection pool key.
        :rtype:  PoolKey
        """
        # Since we mutate the dictionary, make a copy first
        context = request_context.copy()
        context['scheme'] = context['scheme'].lower()
        context['host'] = context['host'].lower()
    
        # These are both dictionaries and need to be transformed into frozensets
        for key in ('headers', '_proxy_headers', '_socks_options'):
            if key in context and context[key] is not None:
                context[key] = frozenset(context[key].items())
    
        # Map the kwargs to the names in the namedtuple - this is necessary since
        # namedtuples can't have fields starting with '_'.
        for key in list(context.keys()):
            context['key_' + key] = context.pop(key)
    
        # Default to ``None`` for keys missing from the context
        for field in key_class._fields:
            if field not in context:
                context[field] = None
    
>       return key_class(**context)
E       TypeError: __new__() got an unexpected keyword argument 'key_assert_fingerprint'

/usr/lib/python3.6/site-packages/urllib3/poolmanager.py:95: TypeError
============================================================================================= pytest-warning summary =============================================================================================
WC1 None pytest_funcarg__httpserver: declaring fixtures using "pytest_funcarg__" prefix is deprecated and scheduled to be removed in pytest 4.0.  Please remove the prefix and use the @pytest.fixture decorator instead.
WC1 None pytest_funcarg__httpsserver: declaring fixtures using "pytest_funcarg__" prefix is deprecated and scheduled to be removed in pytest 4.0.  Please remove the prefix and use the @pytest.fixture decorator instead.
WC1 None pytest_funcarg__smtpserver: declaring fixtures using "pytest_funcarg__" prefix is deprecated and scheduled to be removed in pytest 4.0.  Please remove the prefix and use the @pytest.fixture decorator instead.
===================================================================== 2 failed, 333 passed, 178 skipped, 3 pytest-warnings in 17.25 seconds ======================================================================
make: *** [Makefile:43: test] Error 1
```

After building and installing vdirsyncer with python-urllib3 1.20, I upgrade back to python-urllib3 1.21, but now `vdirsyncer -vdebug sync` fails with the following.
```
debug: Using 1 maximal workers.
debug: Fetching value for password.fetch with command strategy.
Syncing nextcloud_contacts/Contacts
debug: Loaded status for nextcloud_contacts/Contacts
debug: PROPFIND https://foo.dynu.com/nextcloud/remote.php/carddav/addressbooks/Protist/Contacts/
debug: {'User-Agent': 'Mozilla', 'Content-Type': 'application/xml; charset=UTF-8', 'Depth': '1'}
debug: <?xml version="1.0" encoding="utf-8" ?>
debug:             <D:propfind xmlns:D="DAV:">
debug:                 <D:prop>
debug:                     <D:resourcetype/>
debug:                     <D:getcontenttype/>
debug:                     <D:getetag/>
debug:                 </D:prop>
debug:             </D:propfind>
debug:             
debug: Sending request...
error: Unknown error occured for nextcloud_contacts/Contacts: __new__() got an unexpected keyword argument 'key_assert_fingerprint'
error: Use `-vdebug` to see the full traceback.
debug:   File "/usr/lib/python3.6/site-packages/vdirsyncer/cli/tasks.py", line 73, in sync_collection
debug:     partial_sync=pair.partial_sync
debug:   File "/usr/lib/python3.6/site-packages/vdirsyncer/sync.py", line 255, in sync
debug:     b_info.prepare_new_status()
debug:   File "/usr/lib/python3.6/site-packages/vdirsyncer/sync.py", line 150, in prepare_new_status
debug:     for href, etag in self.storage.list():
debug:   File "/usr/lib/python3.6/site-packages/vdirsyncer/storage/dav.py", line 617, in list
debug:     headers=headers)
debug:   File "/usr/lib/python3.6/site-packages/vdirsyncer/storage/dav.py", line 366, in request
debug:     return http.request(method, url, session=self._session, **more)
debug:   File "/usr/lib/python3.6/site-packages/vdirsyncer/http.py", line 162, in request
debug:     r = func(method, url, **kwargs)
debug:   File "/usr/lib/python3.6/site-packages/requests/sessions.py", line 488, in request
debug:     resp = self.send(prep, **send_kwargs)
debug:   File "/usr/lib/python3.6/site-packages/requests/sessions.py", line 609, in send
debug:     r = adapter.send(request, **kwargs)
debug:   File "/usr/lib/python3.6/site-packages/requests/adapters.py", line 390, in send
debug:     conn = self.get_connection(request.url, proxies)
debug:   File "/usr/lib/python3.6/site-packages/requests/adapters.py", line 296, in get_connection
debug:     conn = self.poolmanager.connection_from_url(url)
debug:   File "/usr/lib/python3.6/site-packages/urllib3/poolmanager.py", line 271, in connection_from_url
debug:     pool_kwargs=pool_kwargs)
debug:   File "/usr/lib/python3.6/site-packages/urllib3/poolmanager.py", line 219, in connection_from_host
debug:     return self.connection_from_context(request_context)
debug:   File "/usr/lib/python3.6/site-packages/urllib3/poolmanager.py", line 230, in connection_from_context
debug:     pool_key = pool_key_constructor(request_context)
debug:   File "/usr/lib/python3.6/site-packages/urllib3/poolmanager.py", line 95, in _default_key_normalizer
debug:     return key_class(**context)
error: 1 out of 2 tasks failed.
```

**Comment by untitaker** [comment]
That is a bug in urllib3 which will be fixed by https://github.com/shazow/urllib3/pull/1157

In the meantime please downgrade `python-urllib3`.

**Comment by hobarrera** [comment]
TBH, I think this is a bug in the ArchLinux package for `requests`, since it's their modifications that break this, right?

**Comment by untitaker** [comment]
Yes, but I imagine they will successfully deny it.

On 30 April 2017 01:51:05 GMT+02:00, Hugo Osvaldo Barrera <notifications@github.com> wrote:
>TBH, I think this is a bug in the ArchLinux package for `requests`,
>since it's their modifications that break this, right?
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/vdirsyncer/issues/627#issuecomment-298201715

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by XenGi** [comment]
@hobarrera What modifications do you mean? I can only see the change for the path to cacerts in the package.

https://git.archlinux.org/svntogit/packages.git/tree/trunk?h=packages/python-requests

**Comment by untitaker** [comment]
Normally, requests bundles its own version of urllib3, currently at the version
1.20. So technically you wouldn't notice this problem when installing
vdirsyncer from PyPI, since the latest requests comes with an older urllib3.
The ArchLinux developers unbundled requests though.

**Comment by XenGi** [comment]
Ah, ok. Why would you bundle a dependency? Doesn't make sense to me. Maybe for deploying an app but for an os this seems strange.

**Comment by untitaker** [comment]
The requests developers bundle, ArchLinux undoes this effort.

I find behavior of both parties inappropriate.

On 30 April 2017 15:15:42 GMT+02:00, Ricardo Band <notifications@github.com> wrote:
>Ah, ok. Why would you bundle a dependency? Doesn't make sense to me.
>Maybe for deploying an app but for an os this seems strange.
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/vdirsyncer/issues/627#issuecomment-298231542

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by XenGi** [comment]
Agreed. Arch should take the package as is but requests shouldn't bundle it's dependencies.
Let's just hope urllib fixes the bug. I want my calendar setup back.

**Comment by hobarrera** [comment]
@XenGi They actually unbundle urllib3 (since requests bundles an older version), and alter requests to use the system version (although that one doesn't work).

TBH, I don't know who to blame: the requests' guys for bundling dependencies and not supporting the latest, or the Arch guys for replaceing them for a newer version that fails under some scenarios.

**Comment by XenGi** [comment]
I would blame the urllib guys because from version 1.20 to 1.20 they broke working software.
But why blame anyone, I mean no one gets paid for this. just hand them a beer while they fix it.

**Comment by protist** [comment]
Fixed in [urllib 1.21.1](https://github.com/shazow/urllib3/releases/tag/1.21.1), which just hit the Arch repos. I can confirm that I can build and run vdirsyncer fine now. Thank you for the responses!

**Comment by XenGi** [comment]
Can confirm. Works now.
